File: src/main/java/io/github/chindeaone/collectiontracker/gui/overlays/LotteryOverlay.java
package io.github.chindeaone.collectiontracker.gui.overlays;

import io.github.chindeaone.collectiontracker.config.ConfigAccess;
import io.github.chindeaone.collectiontracker.config.core.Position;
import io.github.chindeaone.collectiontracker.utils.HypixelUtils;
import io.github.chindeaone.collectiontracker.utils.chat.ChatListener;
import io.github.chindeaone.collectiontracker.utils.rendering.RenderUtils;
import io.github.chindeaone.collectiontracker.utils.rendering.TextUtils;
import io.github.chindeaone.collectiontracker.utils.tab.ForagingStatsWidget;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.Font;
import net.minecraft.client.gui.GuiGraphics;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class LotteryOverlay implements AbstractOverlay {

    private final Position position = ConfigAccess.getLotteryPosition();
    private final List<String> lotteryOverlayLines = new ArrayList<>();
    private boolean renderingAllowed  = true;

    @Override
    public String overlayLabel() {
        return "Lottery Overlay";
    }

    @Override
    public Position position() {
        return this.position;
    }

    @Override
    public boolean isEnabled() {
        return ConfigAccess.isLotteryEnabled() && HypixelUtils.isOnSkyblock();
    }

    @Override
    public boolean isRenderingAllowed() {
        return renderingAllowed;
    }

    @Override
    public void setRenderingAllowed(boolean allowed) {
        this.renderingAllowed = allowed;
    }

    @Override
    public void render(GuiGraphics context) {
        if (!isEnabled()) return;
        List<String> lines = getLotteryLines();

        if (lines.isEmpty()) return;

        RenderUtils.drawOverlayFrame(context, position, () ->
                RenderUtils.renderStrings(context, lines)
        );
    }

    @Override
    public void updateDimensions() {
        if (!isEnabled()) return;
        List<String> lines = getLotteryLines();
        if (lines.isEmpty()) return;

        Font fr = Minecraft.getInstance().font;
        int maxW = 0;
        for (String l : lines) maxW = Math.max(maxW, fr.width(l));
        int h = fr.lineHeight * lines.size();

        position.setDimensions(maxW, h);
    }

    private List<String> getLotteryLines() {
        lotteryOverlayLines.clear();
        if (ConfigAccess.isLotteryInForagingIslandsOnly() && ForagingStatsWidget.getCurrentForagingIsland() == null) return Collections.emptyList();

        lotteryOverlayLines.add("§2Lottery: " + ChatListener.getCurrentLotteryBuff());
        lotteryOverlayLines.add(TextUtils.updateTimer());
        return  lotteryOverlayLines;
    }
}

File: src/main/java/io/github/chindeaone/collectiontracker/gui/overlays/LotteryOverlay.kt
package io.github.chindeaone.collectiontracker.gui.overlays

import io.github.chindeaone.collectiontracker.config.core.Position
import io.github.chindeaone.collectiontracker.config.enableLottery
import io.github.chindeaone.collectiontracker.config.lotteryInForagingIslandsOnly
import io.github.chindeaone.collectiontracker.config.lotteryPosition
import io.github.chindeaone.collectiontracker.utils.chat.ChatListener.currentLotteryBuff
import io.github.chindeaone.collectiontracker.utils.world.ForagingMapping.foragingIslands
import io.github.chindeaone.collectiontracker.utils.world.IslandTracker.currentForagingIsland

class LotteryOverlay : AbstractRotatingPerksOverlay() {
    override val overlayLabel: String = "Lottery"

    override val position: Position get() = lotteryPosition

    override val isEnabled: Boolean get() = enableLottery

    override val buffPrefix get() = "§2Lottery"

    override val currentBuff: String get() = currentLotteryBuff

    override val isIslandAllowed: Boolean get() = !lotteryInForagingIslandsOnly || foragingIslands.contains(currentForagingIsland)
}
